# sasquatch的运用



## Background

​         In the era of genome-wide association studies (GWAS) and personalized medicine, predicting the impact of single nucleotide polymorphisms (SNPs) in regulatory elements is an important goal.

​       Sasquatch performs a comprehensive *k*-mer-based analysis of DNase footprints to determine any *k*-mer's potential for protein binding in a specific cell type and how this may be changed by sequence variants. Therefore, Sasquatch uses an unbiased approach, independent of known transcription factor binding sites and motifs. Sasquatch only requires a single DNase-seq data set per cell type, from any genotype, and produces consistent predictions from data generated by different experimental procedures and at different sequence depths.

​    Sasquatch searches for all 5-, 6-, and 7-mers that lie within the DNase I hypersensitive sites (DHSs) present in a specific cell type. It stores their occurrences, piles up the surrounding DNase I cut sites , and produces a footprint profile for each *k*-mer. Importantly, the footprints are normalized for DNase I sequence bias and strand specificity 。

​     





## human(hg19)

### backgorund (naked genomed of human)

### naked genomed proprecess

Updated the previous script, mainly updated the use of bwa's algorithm, making the operation faster (before the script ran for two and a half days, now the script runs for two hours,the test use hg19 genome)

```bash
#!/bin/bash

[ "$#" -gt "2" ] || { echo "Usage:
for pair-end:
ls ./testSet/*_R1_001.fastq.gz | parallel -j 8 nice -n 19 /wrk/yuanzhen/project_bs/pipeline/DNase/DNase_clean_align.sh {} testout/ 6 

for single-end:
ls ./testSet/*.fastq.gz | parallel -j 8 nice -n 19 /wrk/yuanzhen/project_bs/pipeline/DNase/DNase_clean_align.sh {} testout/ 6 

\$1 is R1 filename,
\$2 is output dir,
\$3 is threads per bwa align,
\$4 is genome file (arabidopsis by default)
"; exit 1; }

if [ -z "$4" ]
  then
      genome="/wrk/data_camds/yz_genome_data/aragenome/TAIR10_chr_all.fas"
  else
      genome=$4
fi
#echo "here"; exit;
name_no_suffix=`basename $1 .R1.fastq`
dir=`dirname $1`
outDir=$2
trimmedDir=$outDir/trimmed/
mkdir -p $trimmedDir
mkdir -p $outDir/withDup

set -o nounset
set -o errexit
set -o pipefail

set -x
trim_galore -q 30 --paired --stringency 5 --fastqc --gzip -o $trimmedDir $1 ${dir}/${name_no_suffix}.R2.fastq
bwa mem -t $3 $genome ${trimmedDir}/${name_no_suffix}.R1_val_1.fq.gz ${trimmedDir}/${name_no_suffix}.R2_val_2.fq.gz -o ${trimmedDir}/${name_no_suffix}.sam
samtools view -q 20 -b -S --threads $3  ${trimmedDir}/${name_no_suffix}.sam -o ${trimmedDir}/${name_no_suffix}.bam
samtools sort -n --threads $3  ${trimmedDir}/${name_no_suffix}.bam -o ${outDir}/withDup/${name_no_suffix}.sorted.bam
samtools fixmate --threads $3 -m ${outDir}/withDup/${name_no_suffix}.sorted.bam ${outDir}/withDup/${name_no_suffix}_mate.bam
samtools sort --threads $3  -o ${outDir}/withDup/${name_no_suffix}_positionsort.bam ${outDir}/withDup/${name_no_suffix}_mate.bam
samtools markdup -r --threads $3  ${outDir}/withDup/${name_no_suffix}_positionsort.bam ${outDir}/${name_no_suffix}.sorted_rmdup.bam


#bedtools intersect -abam ${outDir}/${name_no_suffix}.sorted.bam -b ~/DefaultFilesforProc/Genome_related/hg19/hg19_autosomes_human.bed | bedtools intersect -abam stdin -b ~/DefaultFilesforProc/Genome_related/hg19/hg19_blacklist_MergedList.bed -v > ${outDir}/${name_no_suffix}.sorted_autosome_noBlist.bam
rm -rf ${trimmedDir}${name_no_suffix}.R1_val_1.fq.gz ${trimmedDir}${name_no_suffix}.R2_val_2.fq.gz
rm -rf ${trimmedDir}/${name_no_suffix}.sam
rm -rf ${trimmedDir}/${name_no_suffix}.bam
#rm -rf ${outDir}/${name_no_suffix}.sorted.bam
rm -rf ${outDir}/withDup/${name_no_suffix}_mate.bam
rm -rf ${outDir}/withDup/${name_no_suffix}_positionsort.bam

samtools index -@ 30 ${outDir}/withDup/${name_no_suffix}.sorted.bam

samtools index -@ 30 ${outDir}/${name_no_suffix}.sorted_rmdup.bam

samtools idxstats ${outDir}/${name_no_suffix}.sorted_rmdup.bam | cut -f1 | grep -v mitochondria | grep -v chloroplast | xargs samtools view --threads 30 -b ${outDir}/${name_no_suffix}.sorted_rmdup.bam > ${outDir}/${name_no_suffix}.sorted_rmdup_mc.bam

samtools view -H ${outDir}/${name_no_suffix}.sorted_rmdup_mc.bam | sed -e 's/C/c/' | samtools reheader - ${outDir}/${name_no_suffix}.sorted_rmdup_mc.bam > ${outDir}/${name_no_suffix}.sorted_rmdup.bam
```



### footprint for background

```bash
#!/bin/bash

IDTAG="hg19_human_h_ery_1"
#specify if DNaseI or ATAC data ("dnase" or "atac")
DATA_TYPE="DNase"
#type of sequencing ["singleend" / "pairedend"]
SEQ_TYPE="pairedend"

SASQ_PATH=/wrk/yuanzhen/project_bs/snp_latest/sasquatch-master
SCRIPT_DIR=${SASQ_PATH}/data_processing_pipeline/scripts
OUTPUT_DIR=${SASQ_PATH}/background/${IDTAG}_minus_test

# Generated Strand-specific footprint files
FOOTPRINT_FILE_PLUS=${OUTPUT_DIR}/${IDTAG}_Plus.wig
FOOTPRINT_FILE_MINUS=${OUTPUT_DIR}/${IDTAG}_Minus.wig

# Wig border files indicating the first and last position with recorded DNase I cut sites (e.g. generate with ../scripts/get_footprint_wig_borders.pl )
BORDER_FILE_PLUS=${SASQ_PATH}/background/${IDTAG}/footprints/${IDTAG}_ploidy_removed_Plus_merged_wig_borders
BORDER_FILE_MINUS=${SASQ_PATH}/background/${IDTAG}/footprints/${IDTAG}_ploidy_removed_Minus_merged_wig_borders

#specify bam file / naked digest
BAM_FILE=/wrk/data/ref_data_yz/human/Deproteinized_DNA/test/trimmed/back_mem.sorted_rmdup.bam
#genome Build
BUILD='hg19'
#chose chr sizes and ploidy regions to filter
if [ "${BUILD}" == "hg19" ]
then
	REF_GENOME=/wrk/data/genome/yz_genome_data/hg19/hg19.fa
	BIGWIG_CHRSIZES=/wrk/data/genome/yz_genome_data/hg19/hg19_sizes.txt

elif [ "${BUILD}" == "hg18" ]
then
	REF_GENOME="/databank/raw/hg18_full/hg18_full.fa"
	BIGWIG_CHRSIZES='/t1-data1/user/config/bigwig/hg18_sizes.txt'
fi

echo "starting at ..."
date

mkdir -p ${OUTPUT_DIR}

cd ${OUTPUT_DIR}

### -----------------
### Submit Footprints
### -----------------
### TEST ECHO ###
echo "we will do footprint"
echo "OUTPUT_DIR = ${OUTPUT_DIR}"
echo "BAM_FILE = ${BAM_FILE}"
echo "IDTAG = ${IDTAG}"
echo " ============================== "

date

#check if dase or atac and choose appropriate Print_it_All ... file for footprinting wig generation
if [ "${DATA_TYPE}" == "DNase" ]
	then

		#create strand specific footprints
		samtools view ${BAM_FILE} | perl ${SCRIPT_DIR}/Print_It_All_combined.pl --build ${BIGWIG_CHRSIZES} --name ${IDTAG} --type ${SEQ_TYPE} -

	elif [ "${DATA_TYPE}" == "ATAC" ]
		then

		samtools view ${BAM_FILE} | perl ${SCRIPT_DIR}/Print_It_All_combined_atac.pl --build ${BIGWIG_CHRSIZES} --name ${IDTAG} --type ${SEQ_TYPE} -

	else
		echo "DNase or ATAC? DATA_TYPE not correctly specified!"
fi

echo "footprints done ..."
date


```



### first and last position with recorded DNase I cut sites

```bash
perl /wrk/yuanzhen/project_bs/snp_latest/sasquatch-master/data_processing_pipeline/scripts/get_footprint_wig_borders.pl hg19_human_h_ery_1_Plus.wig > hg19_human_h_ery_1_Plus_wig_footprint_borders

perl /wrk/yuanzhen/project_bs/snp_latest/sasquatch-master/data_processing_pipeline/scripts/get_footprint_wig_borders.pl hg19_human_h_ery_1_Minus.wig > hg19_human_h_ery_1_Minus_wig_footprint_borders
```



### Recording *k*-mer-based DNase I cut profiles

kmer_count_1.sh

```bash
#!/bin/bash

IDTAG="hg19_human_h_ery_1"
#specify if DNaseI or ATAC data ("dnase" or "atac")
DATA_TYPE="DNase"
#type of sequencing ["singleend" / "pairedend"]
SEQ_TYPE="pairedend"

SASQ_PATH=/wrk/yuanzhen/project_bs/snp_latest/sasquatch-master
SCRIPT_DIR=${SASQ_PATH}/data_processing_pipeline/scripts
OUTPUT_DIR=${SASQ_PATH}/background/${IDTAG}_minus_test

# Generated Strand-specific footprint files
FOOTPRINT_FILE_PLUS=${OUTPUT_DIR}/${IDTAG}_Plus.wig
FOOTPRINT_FILE_MINUS=${OUTPUT_DIR}/${IDTAG}_Minus.wig

# Wig border files 
BORDER_FILE_PLUS=/wrk/yuanzhen/project_bs/snp_latest/sasquatch-master/background/hg19_human_h_ery_1_minus_test/hg19_human_h_ery_1_Plus_wig_footprint_borders
BORDER_FILE_MINUS=/wrk/yuanzhen/project_bs/snp_latest/sasquatch-master/background/hg19_human_h_ery_1_minus_test/hg19_human_h_ery_1_Minus_wig_footprint_borders

#specify bam file / naked digest
BAM_FILE=/wrk/data/ref_data_yz/human/Deproteinized_DNA/test/trimmed/back_mem.sorted_rmdup.bam
#genome Build
BUILD='hg19'
#chose chr sizes and ploidy regions to filter
if [ "${BUILD}" == "hg19" ]
then
	REF_GENOME=/wrk/data/genome/yz_genome_data/hg19/hg19.fa
	BIGWIG_CHRSIZES=/wrk/data/genome/yz_genome_data/hg19/hg19_sizes.txt

elif [ "${BUILD}" == "hg18" ]
then
	REF_GENOME="/databank/raw/hg18_full/hg18_full.fa"
	BIGWIG_CHRSIZES='/t1-data1/user/config/bigwig/hg18_sizes.txt'
fi

echo "starting at ..."
date

mkdir -p ${OUTPUT_DIR}


###--------------------------
### Submit Footprint Counting
### -------------------------

#START



mkdir -p ${OUTPUT_DIR}/counts
cd ${OUTPUT_DIR}/counts

seq 22 | parallel -j 12 /wrk/yuanzhen/project_bs/snp_latest/kmer_count_2.sh {} ::: ${OUTPUT_DIR} ::: ${FOOTPRINT_FILE_PLUS} ::: ${FOOTPRINT_FILE_MINUS} ::: ${BORDER_FILE_PLUS} ::: ${BORDER_FILE_MINUS} ::: ${IDTAG} ::: ${REF_GENOME}


#seq 22 | parallel -j 12 nice -n 19 /wrk/yuanzhen/project_bs/snp_latest/kmer_count_2.sh {} /wrk/yuanzhen/project_bs/snp_latest/sasquatch-master/background/hg19_human_h_ery_1_minus_test hg19_human_h_ery_1_Plus.wig hg19_human_h_ery_1_Minus.wig hg19_human_h_ery_1_Plus_wig_footprint_borders hg19_human_h_ery_1_Minus_wig_footprint_borders hg19_human_h_ery_1 /wrk/data/genome/yz_genome_data/hg19/hg19.fa
```



kmer_count_2.sh

```bash
#!/bin/bash

    chromosome=$1
    OUTPUT_DIR=$2
    FOOTPRINT_FILE_PLUS=$3
    FOOTPRINT_FILE_MINUS=$4
    BORDER_FILE_PLUS=$5
    BORDER_FILE_MINUS=$6
    IDTAG=$7
    REF_GENOME=$8
    SASQ_PATH=/wrk/yuanzhen/project_bs/snp_latest/sasquatch-master
    SCRIPT_DIR=${SASQ_PATH}/data_processing_pipeline/scripts
    echo "OUTPUT_DIR	${OUTPUT_DIR}"
    echo "FOOTPRINT_FILE_PLUS	${FOOTPRINT_FILE_PLUS}"
    echo "FOOTPRINT_FILE_MINUS	${FOOTPRINT_FILE_MINUS}"
    echo "BORDER_FILE_PLUS	 ${BORDER_FILE_PLUS}"
    echo "BORDER_FILE_MINUS	 ${BORDER_FILE_MINUS}"
    echo "IDTAG	${IDTAG}"
    echo "REF_GENOME   ${REF_GENOME}"
    for i in 5 6 7
  do

  date

  grep -P "chr${chromosome}\s+" ${BORDER_FILE_PLUS} | perl ${SCRIPT_DIR}/count_kmers_naked_hashed.pl /wrk/yuanzhen/download/sasquatch-master/data_processing_pipeline/kmers/All_Possible_Sequences_${i}.txt ${FOOTPRINT_FILE_PLUS} ${i} - ${REF_GENOME} ${OUTPUT_DIR}/counts/kmer_${i}_${IDTAG}_plus_chromosome${chromosome}

  echo "$i plus done"
  date

  grep -P "chr${chromosome}\s+" ${BORDER_FILE_MINUS} | perl ${SCRIPT_DIR}/count_kmers_naked_hashed_minus.pl /wrk/yuanzhen/download/sasquatch-master/data_processing_pipeline/kmers/All_Possible_Sequences_${i}.txt ${FOOTPRINT_FILE_MINUS} ${i} - ${REF_GENOME} ${OUTPUT_DIR}/counts/kmer_${i}_${IDTAG}_minus_chromosome${chromosome}

  echo "$i minus done"
  date

  done
```



### make each chromosome to a file

```bash
##Merge (5,6,7) kmer profiles from naked chromosome profiles
perl ~/project_bs/snp_latest/sasquatch-master/data_processing_pipeline/scripts/merge_naked_chrs.pl kmer_5_hg19_human_h_ery_1_minus_chromosome*  > kmer_5_hg19_human_h_ery_1_minus_merged

perl ~/project_bs/snp_latest/sasquatch-master/data_processing_pipeline/scripts/merge_naked_chrs.pl kmer_5_hg19_human_h_ery_1_plus_chromosome*  > kmer_5_hg19_human_h_ery_1_plus_merged

perl ~/project_bs/snp_latest/sasquatch-master/data_processing_pipeline/scripts/merge_naked_chrs.pl kmer_6_hg19_human_h_ery_1_minus_chromosome*  > kmer_6_hg19_human_h_ery_1_minus_merged

perl ~/project_bs/snp_latest/sasquatch-master/data_processing_pipeline/scripts/merge_naked_chrs.pl kmer_6_hg19_human_h_ery_1_plus_chromosome*  > kmer_6_hg19_human_h_ery_1_plus_merged

perl ~/project_bs/snp_latest/sasquatch-master/data_processing_pipeline/scripts/merge_naked_chrs.pl kmer_7_hg19_human_h_ery_1_minus_chromosome*  > kmer_7_hg19_human_h_ery_1_minus_merged

perl ~/project_bs/snp_latest/sasquatch-master/data_processing_pipeline/scripts/merge_naked_chrs.pl kmer_7_hg19_human_h_ery_1_plus_chromosome*  > kmer_7_hg19_human_h_ery_1_plus_merged

###mv to a new dir
mkdir chromosome_kmer
mv kmer_5_hg19_human_h_ery_1_minus_chromosome* chromosome_kmer/
mv kmer_6_hg19_human_h_ery_1_minus_chromosome* chromosome_kmer/
mv kmer_7_hg19_human_h_ery_1_minus_chromosome* chromosome_kmer/
mv kmer_5_hg19_human_h_ery_1_plus_chromosome* chromosome_kmer/
mv kmer_6_hg19_human_h_ery_1_plus_chromosome* chromosome_kmer/
mv kmer_7_hg19_human_h_ery_1_plus_chromosome* chromosome_kmer/

```

### erythroid

```bash
ls *.R1.fastq | parallel -j 4 nice -n 19 /wrk/yuanzhen/project_bs/pipeline/DNase/DNase_clean_align.sh {} test/ 30 /wrk/data/genome/yz_genome_data/hg19/hg19.fa

~/project_bs/pipeline/DNase/DNase_call_pk.sh /wrk/data/ref_data_yz/human/erythroid/test /wrk/data/ref_data_yz/human/erythroid/test 20

```



total.sh

```bash
#!/bin/bash
## I would make all the bash scripts together
#Set PIPELINE AND OUTPUT Paths ============================================= #
# Full path to your Sasquatch copy
SASQ_PATH=/wrk/yuanzhen/project_bs/snp_latest/sasquatch-master

# path to pipeline runscripts [default = ${SASQ_PATH}/data_processing_pipeline/pipeline]
#PIPE_DIR=${SASQ_PATH}/data_processing_pipeline/pipeline

# path to used pipeline scripts [default = ${SASQ_PATH}/data_processing_pipeline/scripts]
SCRIPT_DIR=${SASQ_PATH}/data_processing_pipeline/scripts

# set output directory [default = ${SASQ_PATH}/data/${ORGANISM}/${DATA_TYPE}/${IDTAG}]



# Set INPUT ================================================================= #
# path to aligned reads bam file
BAM_FILE=/wrk/data/ref_data_yz/human/erythroid/test/SRR4124603.sorted_rmdup.bam
# full path to peak file [.bed .gff or .narrowPeak supported] e.g. for .bed only 3 columns required
REGIONS_FILE=/wrk/data/ref_data_yz/human/erythroid/test/peak_calling/SRR4124603_peaks.narrowPeak
# the perl script handling the region file decides based on the end of the file name (.gff or (.bed or .narroPeak)) in which columns it has to look for the chromosome and start and stop coordinates


# Set PARAMETERS ========================================================== #
# specify organism ["mouse" or "human"]
ORGANISM="human"

# genome Build ["hg18", "hg19", or "mm9"]
BUILD='hg19'

# IDtag to produce output directory and name the files
IDTAG="hg19_human_h_ery_1"

# specify if DNaseI or ATAC data [default = "DNase"]
# ("ATAC" only supported for test purposes)
# you can select ATAC to adjust the fooptint signal for the reportet transposase shifts
# However we do not publish specific ATAC background and do support the use of ATAC data only for testing purposes
DATA_TYPE="DNase"

# select type of sequencing ["singleend" / "pairedend"]
SEQ_TYPE="pairedend"

OUTPUT_DIR=${SASQ_PATH}/${ORGANISM}/${DATA_TYPE}/${IDTAG}

FILTER_PLOIDY_REGIONS=true

# identifier name of the peak file to produce a ploidy filtered peaks [default = `basename ${PEAK_FILE} .bed`]
PEAK_NAME=`basename ${REGIONS_FILE} .narroPeak`
# how to call the ploidy filtered bed file
REGIONS_FILE_PLOIDY_FILTERED=${OUTPUT_DIR}/${PEAK_NAME}_ploidy_filtered.bed

# ============================================================================ #
# Configure Paths to:
#	genomes reference sequence .fa files
#	chr sizes files; two tab separated columns (chrN, size)
#	ploidic regions to exclude (bed format)
# only requires those specified that are atully used
# ============================================================================ #

# hg19
REF_GENOME_hg19=/wrk/data/genome/yz_genome_data/hg19/hg19.fa
BIGWIG_CHRSIZES_hg19=/wrk/data/genome/yz_genome_data/hg19/hg19_sizes.txt
PLOIDY_REGIONS_hg19=/wrk/data/genome/yz_genome_data/hg19/blacklist_ploidy_hg19.bed

# human DNase
PROPENSITY_PLUS_human_dnase=/wrk/yuanzhen/project_bs/snp_latest/sasquatch-master/background/hg19_human_h_ery_1_minus_test/counts/kmer_6_hg19_human_h_ery_1_plus_merged
PROPENSITY_MINUS_human_dnase=/wrk/yuanzhen/project_bs/snp_latest/sasquatch-master/background/hg19_human_h_ery_1_minus_test/counts/kmer_6_hg19_human_h_ery_1_minus_merged

case "${ORGANISM}" in

	human)
		#select ref genome ploidy regions and chromosome sizes
		case "${BUILD}" in

			hg19)
				REF_GENOME=${REF_GENOME_hg19}
				BIGWIG_CHRSIZES=${BIGWIG_CHRSIZES_hg19}
				PLOIDY_REGIONS=${PLOIDY_REGIONS_hg19}
			;;

		esac

		#source of background cut propensities (DNase)
		pnormsource="h_ery_1"
		PROPENSITY_PLUS=${PROPENSITY_PLUS_human_dnase}
		PROPENSITY_MINUS=${PROPENSITY_MINUS_human_dnase}

	;;
esac


if [ "${FILTER_PLOIDY_REGIONS}" = true ];
then
	bedtools intersect -v -a ${REGIONS_FILE} -b ${PLOIDY_REGIONS} > ${REGIONS_FILE_PLOIDY_FILTERED}
else
	${REGIONS_FILE_PLOIDY_FILTERED} = ${PLOIDY_REGIONS}
fi

mkdir -p ${OUTPUT_DIR}/footprints	#make footprint directory
cd ${OUTPUT_DIR}/footprints

### TEST ECHO ###
echo "we will do footprint"
echo "OUTPUT_DIR = ${OUTPUT_DIR}"
echo "BAM_FILE = ${BAM_FILE}"
echo "IDTAG = ${IDTAG}"
echo " ============================== "

date

#check if dase or atac and choose appropriate Print_it_All ... file for footprinting wig generation
if [ "${DATA_TYPE}" == "DNase" ]
	then

		#create strand specific footprints
		samtools view ${BAM_FILE} | perl ${SCRIPT_DIR}/Print_It_All_combined.pl --build ${BIGWIG_CHRSIZES} --name ${IDTAG} --type ${SEQ_TYPE} -

	elif [ "${DATA_TYPE}" == "ATAC" ]
		then

		samtools view ${BAM_FILE} | perl ${SCRIPT_DIR}/Print_It_All_combined_atac.pl --build ${BIGWIG_CHRSIZES} --name ${IDTAG} --type ${SEQ_TYPE} -

	else
		echo "DNase or ATAC? DATA_TYPE not correctly specified!"
fi

echo "footprints done ..."
date

#sleep to ensure footprinting has finsihed properly
sleep 15

echo "we will start kmer count"

COUNTS=${OUTPUT_DIR}/counts
mkdir -p ${COUNTS}

cd ${COUNTS}
date

echo "SASQ_PATH = ${SASQ_PATH}"
echo "OUTPUT_DIR = ${OUTPUT_DIR}"
echo "REGIONS_FILE_PLOIDY_FILTERED = ${REGIONS_FILE_PLOIDY_FILTERED}"
echo "REF_GENOME = ${REF_GENOME}"
echo "IDTAG = ${IDTAG}"
echo "PROPENSITY_PLUS = ${PROPENSITY_PLUS}"
echo "PROPENSITY_MINUS = ${PROPENSITY_MINUS}"
echo "Pnorm Propensities Source = ${pnormsource}"
echo "Count dir = ${COUNTS}"
echo ""

#kmer count
for i in 5 6 
do

	perl ${SCRIPT_DIR}/count_kmers_pnorm.pl ${SASQ_PATH}/data_processing_pipeline/kmers/All_Possible_Sequences_${i}.txt ${REGIONS_FILE_PLOIDY_FILTERED} ${OUTPUT_DIR}/footprints/${IDTAG}_Plus.wig ${i} ${REF_GENOME} ${PROPENSITY_PLUS} >${COUNTS}/kmers_${i}_count_${IDTAG}_pnorm_${pnormsource}_plus.txt


	echo "$i plus done"
	date

	perl ${SCRIPT_DIR}/count_kmers_pnorm_minus.pl ${SASQ_PATH}/data_processing_pipeline/kmers/All_Possible_Sequences_${i}.txt ${REGIONS_FILE_PLOIDY_FILTERED} ${OUTPUT_DIR}/footprints/${IDTAG}_Minus.wig ${i} ${REF_GENOME} ${PROPENSITY_MINUS} >${COUNTS}/kmers_${i}_count_${IDTAG}_pnorm_${pnormsource}_minus.txt

	echo "$i minus done"
	date

done

echo "Total reads: `samtools view ${BAM_FILE} | wc -l`" >${OUTPUT_DIR}/read_stats.txt
echo "Number of Peaks:: `wc -l ${REGIONS_FILE_PLOIDY_FILTERED}`" >>${OUTPUT_DIR}/read_stats.txt
echo "Reads in Peaks:: `bedtools intersect -wa -a ${BAM_FILE} -b ${REGIONS_FILE_PLOIDY_FILTERED} | wc -l`" >>${OUTPUT_DIR}/read_stats.txt
```







## arabidopsis

```bash
/wrk/yuanzhen/project_bs/snp_latest/ara_footprint.sh test.bam test.narrowPeak singleend DNase test/ seedling_leaf
```



```R
source("/wrk/yuanzhen/project_bs/snp_latest/sasquatch-master/R_utility/functions_sasq_r_utility.R")


rf_plot <- function(tissue,kmer,oudir,seqtype="DNase",
                    dataDir="/wrk/yuanzhen/project_bs/snp_latest/ara_control/tair10/DNase",
                    motifdatabase="/var/www/html/sysbiomotif/yuanzhen/arabidopsis/root/motiftotal.Rds"){
my_plots <- vector('list',length(tissue))
  for(i in seq_along(tissue)){
  message(i)
  my_plots[[i]] <- local(
    {
      if(nchar(kmer)==7){
        kmer <- kmer
      }else{
        seq <- kmer
        dl <- DissectSequence(seq, kl=7, list=FALSE)
        dl.sfr <- lapply(dl, function(x){
          x <- GetSFR(kmer=x,tissue=tissue[[i]],data.dir=dataDir,pnorm.tag="ara",vocab.flag=F,frag.type=seqtype)
        }) %>% unlist()
        kmer <- dl[which(dl.sfr==max(dl.sfr))]
      }
      fp <- GetFootprint(kmer=kmer, tissue=tissue[[i]], data.dir=dataDir, pnorm.tag="ara" , frag.type=seqtype, smooth=TRUE)

      sh <- SobelBorders(fp$profile, kl=nchar(kmer))
      sfr <- GetSFR(kmer=kmer,tissue=tissue[[i]],data.dir=dataDir,pnorm.tag="ara",vocab.flag=F,frag.type=seqtype)
      if(i==1){
        p <- PlotSingle(profile=fp$profile,
                        kl=nchar(kmer),
                        plot.shoulders=FALSE,
                        shoulders = sh,
                        ylim=c(0,max(fp$profile)+0.001),
                        xlim = c(-60,60),
                        color = "red")+
          ggtitle(paste0(tissue[[i]]," ","SFR:",round(sfr,3)))+
          theme(plot.title = element_text(size = 10,hjust = 0.5,family="Arial"),
                axis.title.x = element_blank(),
                axis.title.y = element_blank(),
                axis.text.x = element_text(size = 6),
                axis.text.y = element_text(size = 6))+
          labs("kmer")
        print(p)
      }else{
        p <- PlotSingle(profile=fp$profile,
                        kl=nchar(kmer),
                        plot.shoulders=FALSE,
                        shoulders = sh,
                        ylim=c(0,max(fp$profile)+0.001),
                        xlim = c(-60,60),
                        color = "red")+
          ggtitle(paste0(tissue[[i]]," ","SFR:",round(sfr,3)))+
          theme(plot.title = element_text(size = 10,hjust = 0.5,family="Arial"),
                #axis.text.y = element_blank(),
                #axis.ticks.y = element_blank(),
                axis.title.y = element_blank(),
                axis.title.x = element_blank(),
                axis.text.x = element_text(size = 6),
                axis.text.y = element_text(size = 6))
        print(p)

      }
    }
                     )
  }
 #my_pic_name <- stringr::str_c("my_plots","[[",seq_along(tissue),"]]",collapse = ",")
 #eval(parse(text = (str_c("ggpubr::ggarrange(",my_pic_name,",nrow=1,common.legend=T",")",collapse = ""))))
#ggpubr::ggarrange(plotlist = my_plots,nrow = 1,common.legend = T)
png_t <-patchwork::wrap_plots(my_plots,nrow = 1)
pwm_ati <- readr::read_rds(motifdatabase)
ati_consensus <- universalmotif::convert_motifs(pwm_ati) %>% universalmotif::to_df() %>% .$consensus
motif_p <- universalmotif::view_motifs(pwm_ati[[which(ati_consensus==kmer)]],show.positions = F,use.type = "PPM",sort.positions = T)+
  ylab("")+ggtitle(pwm_ati[[which(ati_consensus==kmer)]]@name)+
  theme(axis.ticks.y =element_blank(),
        axis.line.y = element_blank(),
        axis.text.y = element_blank(),
        plot.title = element_text(size = 8,hjust = 0.5,family="Arial"))
plot_sc <- cowplot::ggdraw() + cowplot::draw_plot(png_t, 0, 0, 1, 1) + cowplot::draw_plot(motif_p, 0.8, 0.76, 0.16, 0.16)
ggplot2::ggsave(paste0(oudir,"/",kmer,".png"),plot_sc)
}

pwm_ati <- readr::read_rds("/var/www/html/sysbiomotif/yuanzhen/arabidopsis/root/motiftotal.Rds")
ati_consensus <- universalmotif::convert_motifs(pwm_ati) %>% universalmotif::to_df() %>% .$consensus
lapply(ati_consensus,function(x)rf_plot(c("leaf_jjm","seed_coat","open_flower","rootNOhair_10day"),x,"~/project_bs/snp_latest/ati_motif/"))

```





##  reference

- [Sasquatch: predicting the impact of regulatory SNPs on transcription factor binding from cell- and tissue-specific DNase footprints](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5630036/)

- [main scripts](https://github.com/Hughes-Genome-Group/sasquatch)
